operator-pipeline-images/operatorcert/entrypoints/detect_changed_operators.py

```diff
@@ -78,123 +147,76 @@ def detect_changed_operators(  # pylint: disable=too-many-locals,too-many-branch
-        # Ensure the head commit is checked out (mostly needed for tests)
-        git_repo_head.head.reference = git_repo_head.create_head(  # type: ignore
-            "pr_head", head_commit, force=True
-        )
-        git_repo_head.head.reset(index=True, working_tree=True)
-        # Make a temporary clone of the repo and check out the base commit
-        git_repo_base = git_repo_head.clone(repo_clone_path)
-        git_repo_base.create_head("pr_head", head_commit, force=True)
-        git_repo_base.head.reference = git_repo_base.create_head(  # type: ignore
-            "pr_base", base_commit, force=True
-        )
-        git_repo_base.head.reset(index=True, working_tree=True)
-
-        # Extract all unique names of changed files between pr_base and pr_head, including
-        # previous names for renamed files
-        affected_files = {
-            y
-            for x in git_repo_base.heads.pr_base.commit.diff("pr_head")
-            for y in (x.a_path, x.b_path)
-            if y
-            is not None  # According to the GitPython docs, a_path and b_path can be None
-        }
-
-        LOGGER.debug("Affected files: %s", affected_files)
-
-        non_operator_files: Set[str] = set()
-        all_affected_bundles: Dict[str, Set[Optional[str]]] = {}
-
-        for filename in affected_files:
-            operator_name, bundle_version = _find_owner(filename)
-            if operator_name is None:
-                non_operator_files.add(filename)
+    # Phase 1: using the diff between PR head and base, detect which bundles
+    # have been affected by the PR
+
+    affected_files = github_pr_affected_files(pr_url)
+
+    LOGGER.debug("Affected files: %s", affected_files)
+
+    non_operator_files: Set[str] = set()
+    all_affected_bundles: Dict[str, Set[Optional[str]]] = {}
+
+    for filename in affected_files:
+        operator_name, bundle_version = _find_owner(filename)
+        if operator_name is None:
+            non_operator_files.add(filename)
+        else:
+            all_affected_bundles.setdefault(operator_name, set()).add(
+                # This is None for files within an operator but outside
+                # a bundle (i.e.: ci.yaml)
+                bundle_version
+            )
+
+    LOGGER.debug("Affected bundles: %s", all_affected_bundles)
+    LOGGER.debug("Files not belonging to an operator: %s", non_operator_files)
+
+    # Phase 2: Now that we know which operators and bundles have changed, determine exactly
+    # what was added, modified or removed
+
+    added_operators: Set[str] = set()
+    modified_operators: Set[str] = set()
+    deleted_operators: Set[str] = set()
+    added_bundles: Set[Tuple[str, str]] = set()
+    modified_bundles: Set[Tuple[str, str]] = set()
+    deleted_bundles: Set[Tuple[str, str]] = set()
+
+    head_repo = OperatorRepo(repo_path)
+    base_repo = OperatorRepo(base_path)
+
+    for (  # pylint: disable=too-many-nested-blocks
```

Comment:
Do you think this can be refactored or split into multiple functions that doesn't violate pylint? With a function that long, with so many variables, and nested conditions it is quite hard to read the code.

Reply:
I don't think we can get rid of the nested IFs because that's the core logic of the whole module, but I can try to break the main function in two or three parts to simplify it a bit.

Reply:
That looks really better :+1: